knative-operator/pkg/common/telemetry/telemetry_controller.go

```diff
@@ -0,0 +1,122 @@
+package telemetry
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/apimachinery/pkg/runtime"
+	kafkasourcev1beta1 "knative.dev/eventing-contrib/kafka/source/pkg/apis/sources/v1beta1"
+	eventingsourcesv1beta1 "knative.dev/eventing/pkg/apis/sources/v1beta1"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type op int
+
+const (
+	inc op = iota
+	dec
+)
+
+var typesAndMetrics = []runtime.Object{
+	&servingv1.Service{},
+	&servingv1.Revision{},
+	&servingv1.Route{},
+	&eventingsourcesv1beta1.PingSource{},
+	&eventingsourcesv1beta1.ApiServerSource{},
+}
+
+// Add creates a new Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileResourcesForTelemetry{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("telemetry-resources-discovery-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+	for _, tp := range typesAndMetrics {
+		if err := c.Watch(&source.Kind{Type: tp}, &handler.EnqueueRequestForObject{}, skipUpdatePredicate{}, updateMetricsDeletePredicate{}, updateMetricsCreatePredicate{}); err != nil {
```

Comment:
Seems like we only need the watch functionality? Can we add that in a different file that doesn't look like a controller to make that a bit more obvious?

Reply:
You mean in an existing file?

Reply:
No necessarily, but if we are not going to make this into a controller, let's not make it look like one.